commit a2b6c319bc7195b74256746a3f5d6a42c7039d85
Author: Kalesh Singh <kaleshsingh@google.com>
Date:   Sun Oct 13 13:19:33 2024 -0700

    ltp: syscalls: mlock: Fix fault granularity
    
    Android emulates a userspace 16KiB page size on x86_64 emulators.
    The kernel however remains 4KiB and page faults are therefore handled
    in 4KiB granulatiry.
    
    Fix the mlock tests to account for this case. The changes are
    effectively no-ops for all targets except Android's x86_64 16KiB emulators
    
    Bug: 366098040
    Test: atest vts_ltp_test_x86_64:syscalls.mlock201_64bit#syscalls.mlock201_64bit
    Change-Id: I970e74a07cf314e6506ed07b8d1e096a07c9756d
    Signed-off-by: Kalesh Singh <kaleshsingh@google.com>

diff --git a/include/pgsize_helpers.h b/include/pgsize_helpers.h
index 9c613a5bc..6850f5871 100644
--- a/include/pgsize_helpers.h
+++ b/include/pgsize_helpers.h
@@ -20,12 +20,20 @@ static inline size_t kernel_page_size(void)
 {
     return 0x1000;
 }
-#else
+
+/* Handle upto MAX_PAGE_SIZE emulation on 4KiB kernel base page size system */
+#define DECLARE_MINCORE_VECTOR(vec_name, num_pages) \
+    unsigned char vec_name[(num_pages) * (MAX_PAGE_SIZE / 4096)]
+
+#else	/* !defined(__x86_64__) */
+
 static inline size_t kernel_page_size(void)
 {
 	return getpagesize();
 }
-#endif
+
+#define DECLARE_MINCORE_VECTOR(vec_name, num_pages) unsigned char vec_name[(num_pages)]
+#endif  /* defined(__x86_64__) */
 
 /*
  * NOTE: For all cases except Android x86_64 page size emulators,
@@ -51,4 +59,18 @@ static size_t nr_kernel_pgs_to_nr_pgs(size_t nr_pages)
     return nr_pages / nr_kernel_pages(getpagesize());
 }
 
+/*
+ * Test populating the last partial 4KiB page if the page size is emulated,
+ * instead of the first. This faults in the preceding pages for convenient
+ * test accounting.
+ */
+#define MLOCK_PAGE_SIZE_EMULATION_OFFSET(tcases)                                            \
+do {                                                                                        \
+    if (getpagesize() != kernel_page_size()) {                                              \
+        for (int i = 0; i < ARRAY_SIZE(tcases); i++) {                                      \
+            struct tcase *test = tcases + i;                                                \
+            if (test->offset > 0) test->offset = pgsz - kernel_page_size() + test->offset;  \
+        }                                                                                   \
+    }                                                                                       \
+} while (0)
 #endif /* PGSIZE_HELPER_H */
diff --git a/testcases/kernel/syscalls/mlock2/mlock201.c b/testcases/kernel/syscalls/mlock2/mlock201.c
index b0e5f12b7..1994a119d 100644
--- a/testcases/kernel/syscalls/mlock2/mlock201.c
+++ b/testcases/kernel/syscalls/mlock2/mlock201.c
@@ -25,17 +25,19 @@
 #include "tst_test.h"
 #include "lapi/syscalls.h"
 #include "lapi/mlock2.h"
+#include "pgsize_helpers.h"
 
 #define PAGES	8
 #define HPAGES	(PAGES / 2)
 
 static size_t pgsz;
-static unsigned char vec[PAGES+1];
+
+static DECLARE_MINCORE_VECTOR(vec, PAGES+1);
 
 static struct tcase {
 	size_t populate_pages;
 	size_t lock_pages;
-	size_t offset;
+	ssize_t offset; /* Can be negative */
 	size_t exp_vmlcks;
 	size_t exp_present_pages;
 	int flag;
@@ -78,12 +80,12 @@ static size_t check_locked_pages(char *addr, size_t len, size_t num_pgs)
 
 	SAFE_MINCORE(addr, len, vec);
 
-	for (n = 0; n < num_pgs; n++) {
+	for (n = 0; n < nr_pgs_to_nr_kernel_pgs(num_pgs); n++) {
 		if (vec[n] & 1)
 			act_pages++;
 	}
 
-	return act_pages;
+	return nr_kernel_pgs_to_nr_pgs(act_pages);
 }
 
 static void verify_mlock2(unsigned int n)
@@ -137,6 +139,7 @@ end2:
 static void setup(void)
 {
 	pgsz = getpagesize();
+	MLOCK_PAGE_SIZE_EMULATION_OFFSET(tcases);
 }
 
 static struct tst_test test = {
